### ISSUE TITLE
[Core] Clique engine for Parity Ethereum

### ISSUE THREAD

**Issue by 5chdn** [goerli/testnet#9]
Parity Ethereum is a blockchain client and full-node implementation written in Rust programming language. It supports multiple different chains based on the Ethereum Virtual Machine (EVM) or Parity's WebAssembly VM (pWasm). The client comes with "pluggable consensus," so it has different consensus engines, among others the "Ethash" proof-of-work engine, the "AuRa" proof-of-authority engine, and some developer presets like "Instant Seal" and "Basic Authority."

We want to **add the "Clique" proof-of-authority engine** to that list and allow users to synchronize the Rinkeby, Görli, and Kotti Classic testnets. That breaks down into these major components/tasks:

1. The "Clique" proof-of-authority core engine with proper block validation logic as per [EIP-225](https://github.com/ethereum/EIPs/pull/1570), full chain synchronization, and light/headers verification.
2. The sealing logic to enable the node to be run in "authority mode:" creating, proposing, and sealing new blocks; voting for new validators; unvoting validators to remove them.
3. The "Clique" JSONRPC-APIs that allow users to control the engine via HTTP/WS/IPC APIs.
4. Configurations and chain specifications for the Rinkeby, Görli, and Kotti Classic networks.
5. Tests! Everything should be covered. (This is actually part of all the tasks above, just making sure this is not forgotten about.)

... Profit? All tasks can be claimed via bounties. What is a bounty? That's a fixed amount of money that is claimable by anyone who is submitting a working solution to any of the tasks listed above and sufficient quality. What is sufficient quality? You certainly should have written Rust before and be familiar  with the concepts and best practices of the language. Knowing Parity Ethereum or blockchain experience in general is a bonus. How to claim a bounty? Reach out, name your price and agree on a date for code delivery.

Communication happens on Gitter: [goerli/clique](https://gitter.im/goerli/clique)

**Comment by jwasinger** [comment]
For the parity implementation, I see two milestones that need to be completed before we reach MVP (parity functioning as a signer):
-[] implement voting logic
-[] implement the RPC method for voting

We can create meta-issues to track these but I'd like to break the list down a bit more into granular work tasks for visibility and so that they can be assigned.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 15000.0 DAI (15000.0 USD @ $1.0/DAI)  attached to it as part of the GÃ¶rli Testnet Initiative fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/goerli/testnet/9/2178).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/goerli/testnet/9/2178).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $57,095.27 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by StevenJNPearce** [comment]
Could you not abuse Gitcoin? This is not a bounty. Not clear what it is. Tax evasion?

**Comment by jwasinger** [comment]
@thefallentree and myself have been working on implementing this for months: 
 https://github.com/paritytech/parity-ethereum/pull/9981

**Comment by lilstronghands** [comment]
Curious as to why 5chdn has only approved themselves for this bounty? Why even post the job if you're claiming you're doing it yourself?

**Comment by 5chdn** [comment]
@StevenJNPearce I'm sorry this is not clear what it is. I will improve the description. This is not for previous work. This will be paid out to whoever helps getting the clique engine merged into Parity Ethereum.

**Comment by lazaridiscom** [comment]
> Could you not abuse Gitcoin? This is not a bounty. It's not clear what it is. Pay people for work done previously however you deem fit, but don't post it as a bounty when it has nothing to do with bounties.

@StevenJNPearce  Frustrating, right? Seeing a big bounty, and then realizing that it is already allocated to some devs which were already working on that thing. Been there.

But this happens, e.g. because some funds were delayed or other reasons (e.g. avoid freelance-hire overhead). It's fine I think, for as long as devs (kind of "our colleagues") get payed for their work.

@5chdn , a question to the requirements:

* The final merge to upstream must/should (?) result in one "parity-ethereum" which can handle both, PoW and PoA ?

**Comment by lazaridiscom** [comment]
Edit:

@5chdn , I just realize that it's a "must requirement" (by design). The initial instructions here in the issue are a bit misleading.

I understood:

* implement a `Clique PoA-Engine` (ethereum/EIPs#225) in rustlang 
* use the fork https://github.com/goerli/parity-goerli for your the ongoing work
* submit the final implementation via a PR to upstream paritytech/parity-ethereum#9981

To clarify:

shouldn't upcoming tiny issues be  solved and merged (to upstream) prior to the "big merge" ?

**Comment by 5chdn** [comment]
> Seeing a big bounty, and then realizing that it is already allocated

I'm still accepting bounty hunters, at least one more. We need some help.

> because some funds were delayed

Unfortunately, that's our problem.

> shouldn't upcoming tiny issues be solved and merged (to upstream) prior to the "big merge" ?

I'm open for suggestions. The problem is that this is a massive feature and you cannot implement RPCs before having the engine, and the engine is pretty useless without voting etc. 

Are you interested working on this?

**Comment by lazaridiscom** [comment]
> Are you interested working on this?

Yes, just be aware that it would not make sense to dive into the existent effort.

> We can create meta-issues to track these but I'd like to break the list down a bit more into granular work tasks for visibility and so that they can be assigned.

This was mentioned by @jwasinger in a comment above. It looks that this was not done yet, at least I could not find issues.

-

## Suggestions

After thinking a bit about this, I come up with those tasks (they are in a binding sequence) :

 * A) TASK: 
 * B) TASK: 
 * C) TASK: 

### Task A

**Assess the current situation, provide suggestions / alternatives re processing (mostly "Top Level" stuff)**

You initial commitment would be just TASK A, with a bounty budget of 1K. 

### Task B

**Simplify the implementation of additional engines within parity.**

(this is actually kind of the part "shouldn't upcoming tiny issues be solved and merged (to upstream) prior to the "big merge"? " part of my comment above)

if we go on with TASK B, it would be like "I do my thing autonomously", but at the same time, the devs working currently on the Clique-Implementation can formulate their problems/requirements, and I look to it to provide solution (again: this would be mostly top-level/architectural/ stuff, which is then implemented by the more experienced parity core-devs).

Just as an example, I'd look (far from above, diving in where necessary) into the problem you stated:

> The problem is that this is a massive feature and you cannot implement RPCs before having the engine

I estimate the bounty-volume for Task-B alone at 10K upwards (depending on how much surrounding subsystems will be involved, and how much detail-grade is needed). But the scope can be limited to "what is needed now for the Clique engine" (although I'm afraid its "everything").

### Task C

**Implement the Clique Engine for Parity Ethereum**

Finally, Task C, as a "fall-back" for the case that you current implementation path does not (for any unforeseeable reason) succeed.

Based on Task A and B, i will have a quite clear implementation path/plan. I expect that many existent results can be integrated, and additionally, the gained experience from the devs will ensure that this "2nd attempt" will succeed.

## Clarifications

I hope I expressed things clear. You can contact me additionally via "info . lazaridis . com" if it suits you better.

**Comment by 5chdn** [comment]
Thanks for your suggestions, could you join us on Gitter maybe? https://gitter.im/goerli/clique

**Comment by lazaridiscom** [comment]
Ouh! Me quite allergic to chat, especially in context of "serious analytic talk". Would really prefer asynchronous medium (email, issues, wiki etc.)

For this kind of work I must block most information, and fetch only "as needed" the details. This is quite difficult within chat.

Task A (quite low risk for you) will allow me to further estimate Task B (e.g. if it can be split down into 2 or 3 chunks), so you'll get anyways more info to decide about higher effort/value tasks.

Once you say "go", I expect to have the results in around 3 days.

**Comment by 5chdn** [comment]
"go" ;)

**Comment by lazaridiscom** [comment]
Thanks!

**Comment by thefallentree** [comment]
I ddin't quite follow the discussion here. I mean we already have a working sync + sealing codebase, I'm just in process tidying everything up, so it can be merged. can we try not to over project-manage this further?

**Comment by lazaridiscom** [comment]
@thefallentree , @jwasinger 

To clarify some things:

* I do not want to be the 3rd person in the bounty (I asked for separate funds, and the Task A bounty is a separate fund)
* "We are at 95% means" (in software) sometime "this will never be completed".
* "It works" does not mean (in software) that "it is maintainable".
* If it comes to Task B, then I would negotiate this directly with Parity Technologies (which is the Entity I'm mainly interested in. This does of course not mean that I don't like the Goerli project.).
* as for "over project-manage" one does not have to be 3 decades in IT in order to see that there is "under project management".

The Goerli Project is very(!) difficult. Achieving production-quality whilst using so many independent (not directly controlled) subsystems is a small hell in itself.

Me (Senior Grade, Hardware-Design background, and thus high quality/process standards) just here to help a bit in the overall(!) project.

If you think you don't need help, just let me know, and I'll simply back off.

.

**Comment by thefallentree** [comment]
@lazaridiscom Right now what needs to be done in this issue (clique implementation in parity) is to write the code for clique consensus engine in parity, and I'm just doing that. And when code is working, cleaned up and submit for review, maybe need some help from parity on review and approval.

If you want to discuss other tasks in the overall goerli project, you should collaborate with @5chdn , which he owns. and create other issues tracking that.

The bounty allocation is also entirely 5chdn responsibility, and I have been writing this without any bounty in the first place.  but I, just like @jwasinger , also would like to understand what is the deliverable for task A and curious about that who actually asked for it.

**Comment by lazaridiscom** [comment]
All good, me continuing now, as I'm myself super-curious what Task-A will become.

**Comment by 5chdn** [comment]
> sealing codebase

I wasn't aware, we can run Parity as authority now?

**Comment by thefallentree** [comment]
> > sealing codebase
> 
> I wasn't aware, we can run Parity as authority now?

yep, just pushed my latest code, and made an release

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__The funding of 15000.0 DAI  (15000.0 USD @ $1.0/DAI) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $44,897.26 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by thefallentree** [comment]
@5chdn what's the best way to contact you right now?

**Comment by lazaridiscom** [comment]
@thefallentree , you may try contacting https://twitter.com/MPtherealMVP in goerli related issues.

**Comment by gitcoinbot** [comment]
<table>
<tr>
<td>
<a title='Black Lotus' href='https://gitcoin.co/kudos/1106/black_lotus'> <img width='250' src='https://gitcoin.co/dynamic/kudos/1106/black_lotus' alt='Black Lotus'> </a> 
</td>
<td>
⚡️ A *Black Lotus* Kudos has been sent to @thefallentree for this issue from @5chdn. ⚡️ 

Nice work @thefallentree! 
Your Kudos has automatically been sent in the ETH address we have on file.
</td>
</tr>
</table>

**Comment by gitcoinbot** [comment]
<table>
<tr>
<td>
<a title='Rockstar' href='https://gitcoin.co/kudos/1152/rockstar'> <img width='250' src='https://gitcoin.co/dynamic/kudos/1152/rockstar' alt='Rockstar'> </a> 
</td>
<td>
⚡️ A *Rockstar* Kudos has been sent to @jwasinger for this issue from @5chdn. ⚡️ 

Nice work @jwasinger! 
To redeem your Kudos, login to Gitcoin at https://gitcoin.co/explorer and select 'Claim Kudos' from dropdown menu in the top right, or check your email for a link to the Kudos redemption page. 
</td>
</tr>
</table>
